### PR TITLE
Added subscription index

### PR DESCRIPTION
**Pull request by JanEggers**
the more subscriptions there are the faster the index is compared to linear search

|       Method |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |------------:|----------:|----------:|-------:|------:|------:|----------:|
| LinearSearch | 10,091.1 ns | 171.25 ns | 151.81 ns |      - |     - |     - |         - |
|  IndexSearch |    156.0 ns |   3.10 ns |   4.54 ns | 0.0572 |     - |     - |     241 B |

### CODE REVIEW

File: Source/MQTTnet/Server/MqttSubscriptionIndex.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+﻿using MQTTnet.Protocol;
+using System.Collections.Generic;
+
+namespace MQTTnet.Server
+{
+    public class MqttSubscriptionIndex : Dictionary<string, MqttSubscriptionNode>
```
[the remark this answers is not in the document]
Comment (JanEggers) [reply Source/MQTTnet/Server/MqttSubscriptionIndex.cs]: oki

[the remark this answers is not in the document]
Comment (chkr1011) [reply Source/MQTTnet/Server/MqttSubscriptionIndex.cs]: 👍

**Comment by chkr1011** [comment]
@JanEggers Is this still a draft? Oder is here progress at all? 😄

**Comment by JanEggers** [comment]
if you are ok with the design I will improve test coverage and than it can be merged. I just didnt want to start writing tests to cover all cases before the design is approved.

**Comment by chkr1011** [comment]
For me it looks good. Not over complicated. If this will work for .net 452, UWP etc. (is there probably span used?) I am fine with it.

The only downside I can see that it will allocate RAM right? The other does not if I am wrong.

**Comment by JanEggers** [comment]
>  If this will work for .net 452, UWP etc. 
it does

> (is there probably span used?)
it does not, I really would like to but the problem is span is only stack.  the index lives on the heap. I simply cannot use span as a dictionary key. that would be required to make the lookup allocation free. 

for details:
https://github.com/dotnet/runtime/issues/27229

> The only downside I can see that it will allocate RAM right?
yes but only for splitting the topic into chunks (gen0 very short living) as the performance is way better due to the improved lookup speed id say its worth it

**Comment by chkr1011** [comment]
@JanEggers Is this done? It is still marked as "Draft".

**Comment by JanEggers** [comment]
hi @chkr1011 sry I really have no time..... its basicly about confidence tests. I dont know if you have time for a review, if so go for it.

**Comment by chkr1011** [comment]
No problem 😄 Is it possible to move the branch to this repo? Then I can also work on it and we can delete the pull request as soon as this gets finished.

**Comment by JanEggers** [comment]
> Is it possible to move the branch to this repo?

sure. I guess you could do that, but because I created a pullrequest in your repo you are allowed to push changes on my branch anyway.

**Comment by chkr1011** [comment]
I will close this due to inactivity. But we should continue with this topic any time soon.

**Pull request closed without merging**